### PR TITLE
SYNCT-276: Extend the configuration to give fine grained control

### PR DESCRIPTION
**Pull request by areklalo**
* Added the "preferredClient" field to the ClassConfig and added the Serializable interface to the config classes
* Changed the approach to determine the preferred client (first try to use the class preferred client) and updated the unit tests
* Added support of specific clients configuration

### CODE REVIEW

File: api/src/main/java/org/openmrs/module/sync2/api/service/impl/SyncConfigurationServiceImpl.java [hunk 1]
```diff
@@ -78,4 +100,27 @@ private String getConfigFilePath() {
...
+        }
+        return configuration;
+    }
+
+    private List<ClassConfiguration> getClassesConfiguration(SyncMethodConfiguration methodConfiguration) {
+        List<ClassConfiguration> result = new ArrayList<>();
+        if (methodConfiguration != null) {
+            result = methodConfiguration.getClasses();
+        }
+        return result;
```
Comment (wojtekbu) [inline api/src/main/java/org/openmrs/module/sync2/api/service/impl/SyncConfigurationServiceImpl.java]: maybe use Optional instead a method with if statement

File: api/src/main/java/org/openmrs/module/sync2/api/sync/SyncClient.java [hunk 2]
```diff
@@ -130,38 +131,54 @@ private void prepareRestTemplate(ClientHelper clientHelper) {
...
 		Class<?> clazz = helper.resolveClassByCategory(category);
 
-		return restTemplate.exchange(sendRequest(category, destinationUrl, clientName,
-				new InnerRequest(helper.retrieveRequest(resourceUrl))),
-				clazz).getBody();
+		RequestEntity request = helper.retrieveRequest(resourceUrl);
+		if (!SyncUtils.clientHasSpecificAddress(clientName, instance)) {
+			request = sendRequest(category, destinationUrl, clientName, new InnerRequest(request));
+		}
+		return exchange(request, clazz).getBody();
```
Comment (wojtekbu) [inline api/src/main/java/org/openmrs/module/sync2/api/sync/SyncClient.java]: are you switching between wrapped and unwrapped request?

Comment (wojtekbu) [reply api/src/main/java/org/openmrs/module/sync2/api/sync/SyncClient.java]: Maybe rename clientHasSpecificAddress method or move this if statement to separate method

File: api/src/main/java/org/openmrs/module/sync2/api/sync/SyncClient.java [hunk 3]
```diff
@@ -130,38 +131,54 @@ private void prepareRestTemplate(ClientHelper clientHelper) {
...
-			String clientName) throws RestClientException, URISyntaxException {
+			String clientName, OpenMRSSyncInstance instance) throws RestClientException, URISyntaxException {
 		ClientHelper helper = ClientHelperFactory.createClient(clientName);
-		InnerRequest request = new InnerRequest(helper.createRequest(resourceUrl, object));
 
-		return restTemplate.exchange(sendRequest(category, destinationUrl, clientName, request), String.class);
+		RequestEntity request = helper.createRequest(resourceUrl, object);
+		if (!SyncUtils.clientHasSpecificAddress(clientName, instance)) {
+			request = sendRequest(category, destinationUrl, clientName, new InnerRequest(request));
+		}
```
Comment (wojtekbu) [inline api/src/main/java/org/openmrs/module/sync2/api/sync/SyncClient.java]: as above

File: api/src/main/java/org/openmrs/module/sync2/api/sync/SyncClient.java [hunk 4]
```diff
@@ -130,38 +131,54 @@ private void prepareRestTemplate(ClientHelper clientHelper) {
...
+			String clientName, OpenMRSSyncInstance instance) throws URISyntaxException {
 		ClientHelper helper = ClientHelperFactory.createClient(clientName);
-		InnerRequest request = new InnerRequest(helper.deleteRequest(resourceUrl, uuid));
 
-		return restTemplate.exchange(sendRequest(category, destinationUrl, clientName, request), String.class);
+		RequestEntity request = helper.createRequest(resourceUrl, uuid);
+		if (!SyncUtils.clientHasSpecificAddress(clientName, instance)) {
+			request = sendRequest(category, destinationUrl, clientName, new InnerRequest(request));
+		}
+		return exchange(request, String.class);
```
Comment (wojtekbu) [inline api/src/main/java/org/openmrs/module/sync2/api/sync/SyncClient.java]: as above

File: api/src/main/java/org/openmrs/module/sync2/api/sync/SyncClient.java [hunk 5]
```diff
@@ -130,38 +131,54 @@ private void prepareRestTemplate(ClientHelper clientHelper) {
...
+			String clientName, OpenMRSSyncInstance instance) throws URISyntaxException {
 		ClientHelper helper = ClientHelperFactory.createClient(clientName);
-		InnerRequest request = new InnerRequest(helper.updateRequest(resourceUrl, object));
 
-		return restTemplate.exchange(sendRequest(category, destinationUrl, clientName, request), String.class);
+		RequestEntity request = helper.createRequest(resourceUrl, object);
+		if (!SyncUtils.clientHasSpecificAddress(clientName, instance)) {
+			request = sendRequest(category, destinationUrl, clientName, new InnerRequest(request));
+		}
+		return exchange(request, String.class);
```
Comment (wojtekbu) [inline api/src/main/java/org/openmrs/module/sync2/api/sync/SyncClient.java]: as above

File: api/src/main/java/org/openmrs/module/sync2/api/utils/SyncUtils.java [hunk 6]
```diff
@@ -101,7 +131,13 @@ public static String getFullUrl(String baseAddress, String path) {
 	}
 
 	private static String getPushPath(String pathWithId) {
-		return pathWithId.substring(0, pathWithId.lastIndexOf("/"));
+		String result = null;
+		if (pathWithId.contains("/")) {
+			result = pathWithId.substring(0, pathWithId.lastIndexOf("/"));
```
Comment (wojtekbu) [inline api/src/main/java/org/openmrs/module/sync2/api/utils/SyncUtils.java]: magical number

File: api/src/main/java/org/openmrs/module/sync2/SyncConstants.java [hunk 7]
```diff
@@ -40,6 +40,9 @@
 
     public static final String AUDIT_MESSAGE_VOIDED_FIELD_NAME = "voided";
 
+
+    public static final String ALL_OPERATIONS = "ALL";
```
[the remark this answers is not in the document]
Comment (areklalo) [reply api/src/main/java/org/openmrs/module/sync2/SyncConstants.java]: Corrected, thanks.

File: api/src/main/java/org/openmrs/module/sync2/api/utils/SyncUtils.java [hunk 8]
```diff
@@ -106,7 +107,68 @@ public static String getPullUrl(Map<String, String> resourceLinks, String client
...
+		String login = null;
+		LinkedHashMap<String, ClientConfiguration> availableConfigurations = getSyncConfigurationService()
+				.getSyncConfiguration().getGeneral().getClients();
+		if (!availableConfigurations.isEmpty()) {
+			ClientConfiguration configuration = availableConfigurations.get(clientName);
+			if (configuration != null) {
+				login = configuration.getLogin();
+			}
+		}
+		return login;
```
Comment (wojtekbu) [inline api/src/main/java/org/openmrs/module/sync2/api/utils/SyncUtils.java]: same logic as below

File: api/src/main/java/org/openmrs/module/sync2/api/utils/SyncUtils.java [hunk 9]
```diff
@@ -106,7 +107,68 @@ public static String getPullUrl(Map<String, String> resourceLinks, String client
...
+		String password = null;
+		LinkedHashMap<String, ClientConfiguration> availableConfigurations = getSyncConfigurationService()
+				.getSyncConfiguration().getGeneral().getClients();
+		if (!availableConfigurations.isEmpty()) {
+			ClientConfiguration configuration = availableConfigurations.get(clientName);
+			if (configuration != null) {
+				password = configuration.getPassword();
+			}
+		}
+		return password;
```
Comment (wojtekbu) [inline api/src/main/java/org/openmrs/module/sync2/api/utils/SyncUtils.java]: same logic as above

File: api/src/main/java/org/openmrs/module/sync2/api/utils/SyncUtils.java [hunk 10]
```diff
@@ -69,14 +100,75 @@ public static String getLocalInstanceId() {
...
+		return login;
+	}
+
+	public static String getClientSpecificPassword(String clientName) {
+		String password = null;
+		LinkedHashMap<String, ClientConfiguration> availableConfigurations = getSyncConfigurationService()
+				.getSyncConfiguration().getGeneral().getClients();
+		if (!availableConfigurations.isEmpty()) {
+			ClientConfiguration configuration = availableConfigurations.get(clientName);
+			if (configuration != null) {
```
Comment (wojtekbu) [inline api/src/main/java/org/openmrs/module/sync2/api/utils/SyncUtils.java]: the same logic as above

File: api/src/main/java/org/openmrs/module/sync2/api/utils/SyncUtils.java [hunk 11]
```diff
@@ -69,14 +100,75 @@ public static String getLocalInstanceId() {
...
 		return getFullUrl(base, getPushPath(resourceLinks.get(clientName)));
+	}
+
+	public static String getClientSpecificLogin(String clientName) {
+		String login = null;
+		LinkedHashMap<String, ClientConfiguration> availableConfigurations = getSyncConfigurationService()
+				.getSyncConfiguration().getGeneral().getClients();
+		if (!availableConfigurations.isEmpty()) {
+			ClientConfiguration configuration = availableConfigurations.get(clientName);
+			if (configuration != null) {
```
Comment (wojtekbu) [inline api/src/main/java/org/openmrs/module/sync2/api/utils/SyncUtils.java]: the same logic as below

File: api/src/main/java/org/openmrs/module/sync2/api/utils/SyncUtils.java (hunk 6 above)
Comment (areklalo) [reply api/src/main/java/org/openmrs/module/sync2/api/utils/SyncUtils.java]: Changed to the constant variable.

File: api/src/main/java/org/openmrs/module/sync2/api/sync/SyncClient.java (hunk 2 above)
Comment (areklalo) [reply api/src/main/java/org/openmrs/module/sync2/api/sync/SyncClient.java]: ok

File: api/src/main/java/org/openmrs/module/sync2/api/sync/SyncClient.java (hunk 3 above)
Comment (areklalo) [reply api/src/main/java/org/openmrs/module/sync2/api/sync/SyncClient.java]: Same as above

File: api/src/main/java/org/openmrs/module/sync2/api/sync/SyncClient.java (hunk 4 above)
Comment (areklalo) [reply api/src/main/java/org/openmrs/module/sync2/api/sync/SyncClient.java]: Same as above

File: api/src/main/java/org/openmrs/module/sync2/api/sync/SyncClient.java (hunk 5 above)
Comment (areklalo) [reply api/src/main/java/org/openmrs/module/sync2/api/sync/SyncClient.java]: Same as above

File: api/src/main/java/org/openmrs/module/sync2/api/service/impl/SyncConfigurationServiceImpl.java (hunk 1 above)
Comment (areklalo) [reply api/src/main/java/org/openmrs/module/sync2/api/service/impl/SyncConfigurationServiceImpl.java]: I believe it won't improve complexity in this place.

**Pull request merged**